src/cluster_migrateslots.c

```diff
@@ -0,0 +1,2119 @@
+    if (c->slot_migration_job->state != SLOT_IMPORT_FAILOVER_REQUESTED) {
+        serverLog(LL_WARNING,
+                  "Received CLUSTER SYNCSLOTS FAILOVER-GRANTED from slot "
+                  "migration %s, but we never sent a failover request. Failing "
+                  "migration.",
+                  c->slot_migration_job->description);
+        finishSlotMigrationJob(c->slot_migration_job, SLOT_MIGRATION_JOB_FAILED,
+                               "Unexpected state machine transition");
+        return;
+    }
+    updateSlotMigrationJobState(c->slot_migration_job,
+                                SLOT_IMPORT_FAILOVER_GRANTED);
+    clusterDoBeforeSleep(CLUSTER_TODO_HANDLE_SLOT_MIGRATION);
+}
+
+slotMigrationJob *createSlotImportJob(client *c,
+                                      clusterNode *source_node,
+                                      char *name,
+                                      list *slot_ranges) {
+    slotMigrationJob *job = zcalloc(sizeof(slotMigrationJob));
+    memcpy(job->name, name, CLUSTER_NAMELEN);
+    memcpy(job->nodename, source_node->name, CLUSTER_NAMELEN);
+    job->ctime = server.unixtime;
+    job->last_update = job->ctime;
+    job->last_ack = job->ctime;
+    job->type = SLOT_MIGRATION_IMPORT;
+    job->slot_ranges = slot_ranges;
+    job->slot_ranges_str = representSlotRangeList(slot_ranges);
+    job->state = SLOT_IMPORT_WAIT_ACK;
+    job->client = c;
+    job->client->slot_migration_job = job;
+    job->description = generateSlotMigrationJobDescription(job, source_node);
+
+    /* We treat slot imports like primaries. Primaries are expected to have a
+     * dedicated query buffer and allocated replication data. */
+    initClientReplicationData(job->client);
+    job->client->querybuf = sdsempty();
+
+    /* Mark all the slots as importing in the kvstore */
+    setSlotImportingStateInAllDbs(slot_ranges, 1);
+
+    serverLog(LL_NOTICE, "New slot import job created: %s.", job->description);
+    return job;
+}
+
+/* This function implements the final part of manual slot failovers,
+ * where the target takes over all the slot migration job's hash slots, and
+ * propagates the new configuration. */
+void performSlotImportJobFailover(slotMigrationJob *job) {
+    serverAssert(job->type == SLOT_MIGRATION_IMPORT);
+    /* 1) Force bump the epoch to facilitate propagation. */
+    clusterBumpConfigEpochWithoutConsensus();
+
+    /* 2) Claim all the slots in the slot migration job to myself. */
+    listNode *ln;
+    listIter li;
+    listRewind(job->slot_ranges, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        slotRange *range = ln->value;
+        for (int i = range->start_slot; i <= range->end_slot; i++) {
+            clusterDelSlot(i);
+            clusterAddSlot(server.cluster->myself, i);
+        }
+    }
+
+    /* 3) Update state and save config. */
+    clearCachedClusterSlotsResponse();
+    clusterUpdateState();
+    clusterDoBeforeSleep(CLUSTER_TODO_SAVE_CONFIG);
```

Comment:
I think we should harden this config update
```suggestion
    clusterDoBeforeSleep(CLUSTER_TODO_SAVE_CONFIG | CLUSTER_TODO_FSYNC_CONFIG);
```

Reply:
Sounds good to me